### ISSUE TITLE
When data become large,parition variables can not initialized successfully

### ISSUE THREAD

**Issue by weidong8405347** [tensorflow/tensorflow#15835]
i use tensorflow to distributed trainning models, i use the partition valriables to store an array data, when the data is not so bigger, everything looks ok,but when the array data become larger, when the session initialize, the partition variables can not initialized and the session will wait util time out.

Describe the problem

Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request. the feat_info can initialize successfully, but the adj_info cannot initialized. the adj_info is larger than feat_info

 i use monitored_session or supervisor to do the initialize_variables, when data is not so large, it will initialize successfully, my machine have more than 300GB memory is enough for the larger variable

Have I written custom code
yes
OS Platform and Distribution
linux platform
TensorFlow installed from
N/A
TensorFlow version
r1.4
Bazel version
CUDA/cuDNN version
no gpu
GPU model and memory
no gpu
Exact command to reproduce
N/A

Source code / logs

i use ps_num = 4, worker_num =4 and i also try some other distributed config, like ps_num=1, worker_num=4, the result is the same
source code:
with tf.device(tf.train.replica_device_setter(
worker_device="/job:worker/task:%d" % task_id,
cluster=cluster_spec)):

  feat_info = tf.get_variable("feature_info", (len(id_map),FLAGS.features_column), tf.float32, trainable=False, partitioner=tf.fixed_size_partitioner(num_workers))
  adj_info = tf.get_variable("adj_info", (len(id_map),FLAGS.max_degree), tf.int64, trainable=False, partitioner=tf.fixed_size_partitioner(num_workers))
 
  with tf.device('/job:worker/task:%d' %task_id):
      adj_local = tf.Variable(tf.constant(minibatch.adj, dtype=tf.int64), trainable=False, name="adj_local", collections=[tf.GraphKeys.LOCAL_VARIABLES])
      feat_local = tf.Variable(tf.constant(features, dtype=tf.float32), trainable=False, name="feat_local", collections=[tf.GraphKeys.LOCAL_VARIABLES])
 
  length, begin, end = split_node_by_task(len(id_map), task_id, num_workers)
  adj = tf.nn.embedding_lookup(adj_info, [x for x in range(begin, end)])
  adj = adj_local
  
  feat = tf.nn.embedding_lookup(feat_info, [x for x in range(begin, end)])
  feat = feat_local
log:
2017-12-08 23:54:17.377290: I tensorflow/core/distributed_runtime/master_session.cc:998] Start master session c2b3ba9b700261ba with config:
INFO:tensorflow:Waiting for model to be ready. Ready_for_local_init_op: None, ready: Variables not initialized: adj_info/part_0, adj_info/part_1, adj_info/part_2, adj_info/part_3, adj_info/part_4, adj_info/part_5, adj_info/part_6, adj_info/part_7, adj_info/part_8, adj_info/part_9, adj_info/part_10, adj_info/part_11, adj_info/part_12, adj_info/part_13, adj_info/part_14, adj_info/part_15
2017-12-09 00:00:35.637019: I tensorflow/core/distributed_runtime/master_session.cc:998] Start master session f35fcf332e3908ec with config:
INFO:tensorflow:Waiting for model to be ready. Ready_for_local_init_op: None, ready: Variables not initialized: adj_info/part_0, adj_info/part_1, adj_info/part_2, adj_info/part_3, adj_info/part_4, adj_info/part_5, adj_info/part_6, adj_info/part_7, adj_info/part_8, adj_info/part_9, adj_info/part_10, adj_info/part_11, adj_info/part_12, adj_info/part_13, adj_info/part_14, adj_info/part_15
and it will alway waiting adj_info to initialize

**Comment by cy89** [comment]
@mrry, I'm hestitating to just refer to Stack Overflow because this seems pretty specific. Do you have any advice to offer?

**Comment by mrry** [comment]
There's nothing obviously wrong in the code, but there a lot of unbound values. We'd need a self-contained reproduction to be able to investigate further.

**Comment by weidong8405347** [comment]
@mrry  these proble bother me long time, so please help me to solve it

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by reedwm** [comment]
@weidong8405347 please give a (ideally short) self-contained code sample that we can run to reproduce the problem.

**Comment by weidong8405347** [comment]
with tf.device(tf.train.replica_device_setter(
        worker_device="/job:worker/task:%d" % task_id,
        cluster=cluster_spec)):
       
      feat_info = tf.get_variable("feature_info", (len(id_map),FLAGS.features_column), tf.float32, trainable=False, partitioner=tf.                                    
fixed_size_partitioner(num_workers * 100))
      adj_info = tf.get_variable("adj_info", (len(id_map),FLAGS.max_degree), tf.int64, trainable=False, partitioner=tf.fixed_size_partitioner(num_workers * 100))
       
      with tf.device('/job:worker/task:%d' %task_id):
          adj_local = tf.Variable(tf.constant(minibatch.adj, dtype=tf.int64), trainable=False, name="adj_local", collections=[tf.GraphKeys.LOCAL_VARIABLES])
          feat_local = tf.Variable(tf.constant(features, dtype=tf.float32), trainable=False, name="feat_local", collections=[tf.GraphKeys.LOCAL_VARIABLES])
       
      length, begin, end = split_node_by_task(len(id_map), task_id, num_workers)
      adj = tf.nn.embedding_lookup(adj_info, [x for x in range(begin, end)])
      adj = adj_local
       
      feat = tf.nn.embedding_lookup(feat_info, [x for x in range(begin, end)])
      feat = feat_local            
      grads_and_vars = model.optimizer.compute_gradients(model.loss)
      clipped_grads_and_vars = [(tf.clip_by_value(grad, -5.0, 5.0) if grad is not None else None, var)
                                for grad, var in grads_and_vars]
      rep_op = tf.train.SyncReplicasOptimizer(model.optimizer,
                                              replicas_to_aggregate=num_workers,
                                              total_num_replicas=num_workers,
                                              use_locking=True)
      tf.logging.info('[%s] Build train op ...' % datetime.now())
      summaries = tf.get_collection(tf.GraphKeys.SUMMARIES)
                             
      apply_gradient_op = rep_op.apply_gradients(clipped_grads_and_vars, global_step=global_step)                                                                      
      init_token_op = rep_op.get_init_tokens_op()
      chief_queue_runner = rep_op.get_chief_queue_runner()
                             
      saver = tf.train.Saver(max_to_keep=FLAGS.max_to_keep)
                             
      summary_op = tf.summary.merge(summaries)
      init_op = tf.global_variables_initializer()
      local_init_op = tf.local_variables_initializer()
      table_op = tf.tables_initializer()
      merge_op = tf.group(init_op, table_op)
    tf.logging.info('[%s] Build Supervisor ...' % datetime.now())
    sv = tf.train.Supervisor(is_chief=is_chief,
                             logdir=FLAGS.checkpoint,
                             init_op=merge_op,
                             local_init_op=local_init_op,
                             summary_op=None,
                             saver=saver,
                             global_step=global_step,
                             recovery_wait_secs=FLAGS.recovery_wait_secs,
                             save_summaries_secs=None,
                             save_model_secs=None)                                                                                                                     
                              
    sess_config = tf.ConfigProto(
        allow_soft_placement=False,
        log_device_placement=FLAGS.log_device_placement)
    if is_chief:              
        tf.logging.info('[%s] Worker %d, Initialize session...' %
                        (datetime.now(), task_id))
    else:                     
        tf.logging.info('[%s] Worker %d, Wait for session to be initialized..' %
                        (datetime.now(), task_id))
                              
    with sv.prepare_or_wait_for_session(target, config=sess_config, max_wait_secs=FLAGS.max_wait_secs) as sess:

**Comment by reedwm** [comment]
@weidong8405347, the example is not completely self contained, since you have some unset variables like `cluster_spec`. Can you give a self-contained example?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by automateljw** [comment]
I meet the same problem, when variable is large and in distribution environment.  chief worker can't initializer the large variable correctly, so i guess it is bug!
the large variable i define is:
```
### problem code
w0 = tf.get_variable('w_0', dtype=dtype, initializer=tf.random_uniform([100000, 256, 32], minval=minval, maxval=maxval, dtype=dtype)) 

### correct code （ just smaller first dimention size）
w0 = tf.get_variable('w_0', dtype=dtype, initializer=tf.random_uniform([30000, 256, 32], minval=minval, maxval=maxval, dtype=dtype)) 
```
I use tf.estimator.train_and_evaluate api to do distributed training, and define w0 in the model_fn.  when i define w0 using[100000, 256, 32]  , I can run the code in the single machine( local mode) correctly, but Error ：
```
### in chief
INFO:tensorflow:Create CheckpointSaverHook.
2018-03-02 12:53:31.539989: I tensorflow/core/distributed_runtime/master_session.cc:1017] Start master session 001a93d5f7256838 with config: device_count { key: "GPU" value: 0 }
[[[[  blocked here ]]]]]

### in worker 0
INFO:tensorflow:Waiting for model to be ready.  Ready_for_local_init_op:  Variables not initialized: w_0, ready: None
```
when i only change w0 from [100000, 256, 32] to [30000, 256, 32], I can run the code correctly both in local mode and distribution mode. So large variable can't initialize correctly in distribution mode

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by weidong8405347** [comment]
still a bug

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shivaniag** [comment]
@weidong8405347 is this still a problem? apologize for delayed response here.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by harshini-gadige** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by u10116032** [comment]
I ran Tensorflow Object Detection project with fasterrcnn_resnet50 on distributed training. All I modified is to add a convLSTM layer after feature extractor.
Recently I want to run on distributed machines through legacy/train.py, and I faced same problem like this issue.

My cluster spec is as below:
```
{
  "cluster": {
    "ps": [
      "localhost:3000"
    ],
    "master": [
      "localhost:3001"
    ],
    "worker": [
      "localhost:3002"
    ]
  },
  "task": {
    "type": "ps/master/worker", 
    "index": 0
  }
}
``` 

For *ps* machine, I set task.type in cluster spec as *ps*, and it runs well.
For *master* machine, I set task.type in cluster spec as *master*, and it runs well too.
It starts training since the default setting of synchronize training in config file is false.
However, for *worker* machine, I set task.type as *worker*. It repeatedly shows 
```
INFO:tensorflow:Waiting for model to be ready.  Ready_for_local_init_op:  None, ready: Variables not initialized: rnn/convlstm_cell/kernel, rnn/convlstm_cell/biases, rnn/convlstm_cell/kernel/Momentum, rnn/convlstm_cell/biases/Momentum
```

I'm wondering if it's a bug for slim.learning.train / supervisor for legacy/trainer.py or not,
should I change to MonitoredTrainingSession ?

**Comment by zhtea** [comment]
meet the same problem too

**Comment by AtekiRyu** [comment]
Hi i meet the same problem with tf1.12 CPU version on Windows.
When create a PartitionedVariable with        
`
ps_count = 5
embeddings = tf.get_variable(
                    "embeddings", [term_hash_bucket_size, embedding_count], dtype=tf.float32,
                    initializer=variable_initializer, partitioner=tf.fixed_size_partitioner(ps_count))
`
it will cost 20mins for sees.run(tf.global_variable_initializer())
When the ps_count=20, it will cost more than 80mins.
